ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/SynchronousResponse.scala

```diff
@@ -29,59 +29,73 @@ import scala.concurrent.{ExecutionContext, Future, TimeoutException}
 class SynchronousResponse[Input, Entry, AcceptedEntry](
     strategy: SynchronousResponse.Strategy[Input, Entry, AcceptedEntry],
     timeToLive: Duration,
+)(implicit
+    executionContext: ExecutionContext,
+    materializer: Materializer,
+    loggingContext: LoggingContext,
 ) {
 
   private val logger = ContextualizedLogger.get(getClass)
 
-  def submitAndWait(submissionId: Ref.SubmissionId, input: Input)(implicit
-      telemetryContext: TelemetryContext,
-      executionContext: ExecutionContext,
-      materializer: Materializer,
+  private def errorLogger(loggingContext: LoggingContext, submissionId: Ref.SubmissionId) =
+    new DamlContextualizedErrorLogger(logger, loggingContext, Some(submissionId))
+
+  private def toGrpcError(
       loggingContext: LoggingContext,
+      submissionId: Ref.SubmissionId,
+  ): PartialFunction[Throwable, Future[Nothing]] = {
+    case _: TimeoutException =>
+      Future.failed(
+        LedgerApiErrors.RequestTimeOut
+          .Reject("Request timed out", definiteAnswer = false)(
+            errorLogger(loggingContext, submissionId)
+          )
+          .asGrpcError
+      )
+    case _: NoSuchElementException =>
+      Future.failed(
+        LedgerApiErrors.ServiceNotRunning
+          .Reject("Party submission")(errorLogger(loggingContext, submissionId))
+          .asGrpcError
+      )
+  }
+
+  private def acknowledged(
+      submissionId: Ref.SubmissionId,
+      ledgerEndBeforeRequest: Option[LedgerOffset.Absolute],
+  ) = {
+    val isAccepted = new Accepted(strategy.accept(submissionId))
+    val isRejected = new Rejected(strategy.reject(submissionId))
+    strategy
+      .entries(ledgerEndBeforeRequest)
+      .collect {
+        case isAccepted(entry) => Future.successful(entry)
+        case isRejected(exception) => Future.failed(exception)
+      }
+      .completionTimeout(FiniteDuration(timeToLive.toMillis, TimeUnit.MILLISECONDS))
+      .runWith(Sink.head)
+      .recoverWith(toGrpcError(loggingContext, submissionId))
+      .flatten
+  }
+
+  private def toResult(
+      submissionId: Ref.SubmissionId,
+      ledgerEndBeforeRequest: Option[LedgerOffset.Absolute],
+      submissionResult: SubmissionResult,
+  ) = submissionResult match {
+    case SubmissionResult.Acknowledged =>
+      acknowledged(submissionId, ledgerEndBeforeRequest)
+    case synchronousError: SubmissionResult.SynchronousError =>
+      Future.failed(synchronousError.exception)
+  }
+
+  def submitAndWait(submissionId: Ref.SubmissionId, input: Input)(implicit
```

Comment:
Can we keep the public function at the beginning? it's a matter of preference, but imho it's more logical like this assuming that you read the code top-down in the dispatch order.

Reply:
interesting, i always preferred public functions in the very end :D 

but dont see it as big concern, so pushing the fix